File: src/main/java/com/aws/emr/spark/cdc/SparkIncrementalPipeline.java
package com.aws.emr.spark.cdc;

import com.aws.emr.common.JobConfig;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.TimeoutException;
import org.apache.iceberg.Snapshot;
import org.apache.iceberg.Table;
import org.apache.iceberg.spark.CommitMetadata;
import org.apache.iceberg.spark.Spark3Util;
import org.apache.iceberg.util.SnapshotUtil;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
import org.apache.spark.sql.catalyst.parser.ParseException;

/**
 * An example of the Iceberg incremental processing pattern applied to CDC: instead of re-scanning
 * the whole changelog on every run, we read only the snapshots appended to
 * {@code accounts_changelog} since the last processed snapshot and MERGE them into
 * {@code accounts_mirror}. The last processed source snapshot id is stored as a commit property
 * ({@code watermark:accounts_changelog}) on the target table, so the read range and the merge are
 * committed atomically and the pipeline is restartable.
 *
 * <p>Both tables are Iceberg format-version 3 (v3) tables. The Spark session, catalog and run
 * environment are selected through {@link JobConfig} {@code key=value} arguments; see
 * {@link JobConfig#usage()}.
 *
 * @author acmanjon@amazon.com
 */
public class SparkIncrementalPipeline {

  private static final Logger log = LogManager.getLogger(SparkIncrementalPipeline.class);

  /** Commit property used to track the last processed source snapshot on the target table. */
  private static final String WATERMARK_KEY = "watermark:accounts_changelog";

  public static void main(String[] args)
      throws NoSuchTableException, ParseException, IOException, TimeoutException {

    JobConfig cfg = JobConfig.fromArgs(args);
    SparkSession spark = cfg.buildSession("CDCIncrementalPipeline");

    spark.sql("CREATE DATABASE IF NOT EXISTS " + JobConfig.DATABASE);
    spark.sql("USE " + JobConfig.DATABASE);

    // The source changelog table (populated by SparkLogChange) ...
    spark.sql(
        """
                        CREATE TABLE IF NOT EXISTS accounts_changelog
                              (
                              operation string,
                              account_id bigint,
                              balance bigint,
                              last_updated timestamp,
                              seq bigint            -- source sequence (LSN surrogate) for deterministic ordering
                              )
                              PARTITIONED BY (days(last_updated),bucket(8, account_id))
                              TBLPROPERTIES (
                                        'table_type'='ICEBERG',
                                        'format-version'='3',
                                        'write.parquet.compression-codec'='zstd',
                                        'compatibility.snapshot-id-inheritance.enabled'='true' );
                        """);
    // ... and the target mirror table where the deduplicated, merged state lives.
    spark.sql(
        """
                        CREATE TABLE IF NOT EXISTS accounts_mirror
                              (account_id bigint,
                              balance float,
                              last_updated timestamp,
                              seq bigint            -- last applied source sequence, for stale-change guards
                              )
                              PARTITIONED BY (bucket(8, account_id))
                              TBLPROPERTIES (
                                        'table_type'='ICEBERG',
                                        'format-version'='3',
                                        'write.delete.mode'='merge-on-read',
                                        'write.update.mode'='merge-on-read',
                                        'write.merge.mode'='merge-on-read',
                                        'write.parquet.compression-codec'='zstd',
                                        'compatibility.snapshot-id-inheritance.enabled'='true' );
                        """);

    // Load the target table and recover the last processed source snapshot id from its history.
    Table mirrorTable = Spark3Util.loadIcebergTable(spark, "accounts_mirror");
    String lastProcessedId = null;
    for (Snapshot snap : SnapshotUtil.currentAncestors(mirrorTable)) {
      lastProcessedId = snap.summary().get(WATERMARK_KEY);
      if (lastProcessedId != null) {
        break;
      }
    }

    // Load the source table and figure out the snapshot we want to process up to.
    Table logSourceTable = Spark3Util.loadIcebergTable(spark, "accounts_changelog");
    if (logSourceTable.currentSnapshot() == null) {
      log.warn("Source table accounts_changelog has no snapshots yet, nothing to process.");
      return;
    }
    String toProcessId = Long.toString(logSourceTable.currentSnapshot().snapshotId());

    if (toProcessId.equals(lastProcessedId)) {
      log.warn("No new snapshots in accounts_changelog since {}, nothing to process.", lastProcessedId);
      return;
    }
    log.warn("Last processed source snapshot was {} and we will process up to {}", lastProcessedId, toProcessId);

    // Incremental read of only the new appends. On the very first run (no watermark yet) we read the
    // whole source table up to the current snapshot.
    Dataset<Row> newLogs;
    if (lastProcessedId == null) {
      newLogs = spark.read().format("iceberg").option("end-snapshot-id", toProcessId).table("accounts_changelog");
    } else {
      newLogs =
          spark
              .read()
              .format("iceberg")
              .option("start-snapshot-id", lastProcessedId) // exclusive
              .option("end-snapshot-id", toProcessId)
              .table("accounts_changelog");
    }
    newLogs.createOrReplaceTempView("accounts_source");

    // Update the target table and record the watermark in the same commit. If the merge fails, the
    // watermark is not advanced and the range is safely reprocessed on the next run.
    CommitMetadata.withCommitProperties(
        Map.of(WATERMARK_KEY, toProcessId),
        () -> {
          spark.sql(CdcSql.mirrorMerge("accounts_mirror", "accounts_source"));
          return 0;
        },
        RuntimeException.class);

    log.warn("Incremental merge complete, watermark advanced to {}", toProcessId);

    // Rollback & replay, if you ever need to reprocess a range:
    // CALL system.rollback_to_snapshot('accounts_mirror', <LAST-CORRECT-SNAPSHOT-ID>);
  }
}

File: python/src/iceberg_streaming/cdc/spark_incremental_pipeline.py
"""CDC incremental processing pipeline.

PySpark counterpart of ``com.aws.emr.spark.cdc.SparkIncrementalPipeline``. Reads only the snapshots
appended to ``accounts_changelog`` since the last processed one and merges them into
``accounts_mirror``.

**Difference from the Java version:** the Java job attaches the source watermark to the *same* Iceberg
commit as the MERGE using the JVM-only ``org.apache.iceberg.spark.CommitMetadata`` thread-local. That
API is impractical to drive from Python workers, so this version stores the watermark as a table
property (``watermark:accounts_changelog``) on ``accounts_mirror`` in a follow-up ``ALTER TABLE``
statement. It is therefore not committed atomically with the MERGE; if the pipeline dies between the
two, the range is simply reprocessed on the next run (the MERGE is idempotent).
"""

from __future__ import annotations

import logging
import sys

from iceberg_streaming.cdc import _sql
from iceberg_streaming.cdc._sql import mirror_merge
from iceberg_streaming.common import DATABASE, JobConfig, Mode

log = logging.getLogger("iceberg_streaming.cdc.spark_incremental_pipeline")

WATERMARK_KEY = "watermark:accounts_changelog"

_MERGE = mirror_merge("accounts_mirror", "accounts_source")


def _read_watermark(spark) -> str | None:
    try:
        rows = spark.sql(f"SHOW TBLPROPERTIES accounts_mirror ('{WATERMARK_KEY}')").collect()
    except Exception:  # property not set / not supported yet
        return None
    if not rows:
        return None
    value = rows[0]["value"]
    # Spark returns a placeholder string when the property is absent.
    if not value or "does not have property" in value:
        return None
    return value


def _current_source_snapshot(spark) -> str | None:
    rows = spark.sql(
        "SELECT snapshot_id FROM accounts_changelog.snapshots ORDER BY committed_at DESC LIMIT 1"
    ).collect()
    return str(rows[0]["snapshot_id"]) if rows else None


def main(argv: list[str] | None = None) -> None:
    cfg = JobConfig.from_args(argv if argv is not None else sys.argv[1:])
    spark = cfg.build_session("PySparkCDCIncrementalPipeline")

    spark.sql(f"CREATE DATABASE IF NOT EXISTS {DATABASE}")
    spark.sql(f"USE {DATABASE}")
    spark.sql(
        cfg.create_table_ddl(
            "accounts_changelog", _sql.CHANGELOG_COLUMNS_DDL, _sql.CHANGELOG_PARTITION_DDL, Mode.COW
        )
    )
    # Merge-on-read mirror by default so the MERGE's deletes become deletion vectors on v3.
    spark.sql(
        cfg.create_table_ddl(
            "accounts_mirror", _sql.MIRROR_COLUMNS_DDL, _sql.MIRROR_PARTITION_DDL, Mode.MOR
        )
    )

    to_process = _current_source_snapshot(spark)
    if to_process is None:
        log.warning("Source table accounts_changelog has no snapshots yet, nothing to process.")
        return

    last_processed = _read_watermark(spark)
    if to_process == last_processed:
        log.warning("No new snapshots in accounts_changelog since %s, nothing to process.", last_processed)
        return
    log.warning("Last processed source snapshot was %s, processing up to %s", last_processed, to_process)

    reader = spark.read.format("iceberg").option("end-snapshot-id", to_process)
    if last_processed is not None:
        reader = reader.option("start-snapshot-id", last_processed)  # exclusive
    reader.table("accounts_changelog").createOrReplaceTempView("accounts_source")

    spark.sql(_MERGE)

    # Persist the watermark (see the module docstring for the atomicity caveat vs. the Java version).
    spark.sql(
        f"ALTER TABLE accounts_mirror SET TBLPROPERTIES ('{WATERMARK_KEY}' = '{to_process}')"
    )
    log.warning("Incremental merge complete, watermark advanced to %s", to_process)


if __name__ == "__main__":
    main()
